sycl/source/detail/graph_impl.hpp

```diff
@@ -197,6 +205,63 @@ class graph_impl {
       : MContext(SyclContext), MDevice(SyclDevice), MRecordingQueues(),
         MEventsMap() {}
 
+private:
+  /// A cache of pointers to exit nodes
+  ///
+  /// This is not used (yet), but depth computation starts from exit nodes
+  /// Perhaps, it might be better to do the exec_order_recompute traversal
+  /// starting from each exit node and working upwards using MPredecessors
+  /// rather than from each root node and doing depth-first to exit nodes?
+  std::vector<node_impl *> MExitNodes;
+
+  /// A sorted multimap capturing the optimal execution/submission order
+  ///
+  /// The SortKey is the depth in the graph for the node_impl in the value
+  /// Depth is the length of the longest dependence chain to any root node
+  std::multimap<int, std::shared_ptr<node_impl>> MExecOrder;
```

Comment:
Preference for saying "breadth-first" rather than optimal, since it could depend on the combination of application & platform what is optimal.
```suggestion
  /// A sorted multimap capturing the breadth-first execution/submission order.
  ///
  /// The SortKey is the depth in the graph for the node_impl in the value.
  /// Depth is the length of the longest dependence chain to any root node.
  std::multimap<int, std::shared_ptr<node_impl>> MExecOrder;
```